src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/adaptive/AdaptiveJqlMemory2.kt

```diff
@@ -0,0 +1,68 @@
+package com.atlassian.performance.tools.jiraactions.api.memories.adaptive
+
+import com.atlassian.performance.tools.jiraactions.api.SeededRandom
+import com.atlassian.performance.tools.jiraactions.api.jql.Jql
+import com.atlassian.performance.tools.jiraactions.api.jql.JqlSupplier
+import com.atlassian.performance.tools.jiraactions.api.memories.JqlMemory
+import com.atlassian.performance.tools.jiraactions.api.memories.JqlMemory2
+import com.atlassian.performance.tools.jiraactions.api.memories.adaptive.jql.JqlPrescription
+import com.atlassian.performance.tools.jiraactions.api.memories.adaptive.jql.JqlPrescriptions
+import com.atlassian.performance.tools.jiraactions.api.page.IssuePage
+import com.atlassian.performance.tools.jiraactions.jql.BuiltInJQL
+import com.atlassian.performance.tools.jiraactions.jql.JqlContextImpl
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import java.util.function.Predicate
+
+class AdaptiveJqlMemory2(
+    private val random: SeededRandom
+) : JqlMemory2 {
+
+    private val logger: Logger = LogManager.getLogger(this::class.java)
+
+    companion object {
+        fun getBuiltInJqlSuppliers(): List<JqlSupplier> = BuiltInJQL.values().toList()
+    }
+
+    private val preparedQueries = mutableListOf<Jql>();
+
+    init {
+        BuiltInJQL.RESOLVED.get(JqlContextImpl(random))?.let { preparedQueries.add(it) }
+        BuiltInJQL.GENERIC_WIDE.get(JqlContextImpl(random))?.let { preparedQueries.add(it) }
+    }
+
+    private val availableJqlSuppliers = mutableSetOf(
```

Comment:
`Memory` impls are implicitly allowed to be not thread-safe, because each VU has its own memories.
It's a nice opportunity to make that assumption explicit and mark `AdaptiveJqlMemory2` as `@NotThreadSafe`.